internal/controller/metalmachine_controller.go

```diff
@@ -191,14 +202,151 @@ func (r *MetalMachineReconciler) reconcileNormal(_ context.Context, machineScope
+		machineScope.Error(err, "failed to patch the MetalMachine with providerid")
+		return ctrl.Result{}, err
+	}
 
-	// Get or create the ServerClaim.
-	// TBD
-	machineScope.Info("Creating ServerClaim", "claim", machineScope.MetalMachine.Name)
+	if serverClaim.Status.Phase != metalv1alpha1.PhaseBound {
+		machineScope.Info("Waiting for ServerClaim to be Bound")
+		return ctrl.Result{
+			RequeueAfter: infrav1alpha1.DefaultReconcilerRequeue,
+		}, nil
+	}
 
 	machineScope.SetReady()
 	machineScope.Logger.Info("MetalMachine is ready")
 
 	return reconcile.Result{}, nil
 }
+
+func (r *MetalMachineReconciler) applyIgnitionSecret(ctx context.Context, log *logr.Logger, capidatasecret *corev1.Secret) (*corev1.Secret, error) {
+	secretObj := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      fmt.Sprintf("ignition-%s", capidatasecret.Name),
+			Namespace: capidatasecret.Namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: corev1.SchemeGroupVersion.String(),
+		},
+		Data: map[string][]byte{
+			// TODO: Make Metal Specific changes in the Ignition if necessary
+			DefaultIgnitionSecretKeyName: capidatasecret.Data["value"],
+		},
+	}
+
+	if err := controllerutil.SetControllerReference(capidatasecret, secretObj, r.Client.Scheme()); err != nil {
+		return nil, fmt.Errorf("failed to set ControllerReference: %w", err)
+	}
+
+	opResult, err := controllerutil.CreateOrPatch(ctx, r.Client, secretObj, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create or patch the IgnitionSecret: %w", err)
+	}
+	log.Info("Created or Patched IgnitionSecret", "IgnitionSecret", secretObj.Name, "Operation", opResult)
+
+	return secretObj, nil
+}
+
+func (r *MetalMachineReconciler) applyServerClaim(ctx context.Context, log *logr.Logger, metalmachine *infrav1alpha1.MetalMachine, ignitionsecret *corev1.Secret) (*metalv1alpha1.ServerClaim, error) {
+	serverClaimObj := &metalv1alpha1.ServerClaim{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      metalmachine.Name,
+			Namespace: metalmachine.Namespace,
+		},
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: metalv1alpha1.GroupVersion.String(),
+			Kind:       "ServerClaim",
+		},
+		Spec: metalv1alpha1.ServerClaimSpec{
+			Power: metalv1alpha1.PowerOn,
+			IgnitionSecretRef: &corev1.LocalObjectReference{
+				Name: ignitionsecret.Name,
+			},
+			// TODO: Allow configuring os-image.
+			Image: "ghcr.io/ironcore-dev/os-images/gardenlinux:1443.3",
+		},
+	}
+
+	// TODO: Define proper contract for ServerSelectors.
+	serverSelector := metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			"clusterapi-workload": "",
+		},
+	}
+	if _, exists := metalmachine.Labels["cluster.x-k8s.io/control-plane"]; exists {
+		serverSelector.MatchLabels = map[string]string{
+			"clusterapi-controlplane": "",
+		}
+	}
+	serverClaimObj.Spec.ServerSelector = &serverSelector
+
+	if err := controllerutil.SetControllerReference(metalmachine, serverClaimObj, r.Client.Scheme(), nil); err != nil {
```

Comment:
shall we move this after `CreateOrPatch` (eg line 325) ? Otherwise we are trying to set reference on claim that does not exist yet

Reply:
Yeah for sure, I will make that change tomorrow, thanks !!